---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
------------------------------------------------------------------------------------------------

Scientific Question: {2 points} What is the core genome of life? Comparative genomics of plant, bacterial and animal sequences.

In its essence, life is but beads on a string, more strictly speaking, nucleic acids on a sugar string. Together, these rearrange into all the marvel of the living world. However, if these variations are what makes everything different, how do we all share this metabolic state which we call life? What do we have in common with a tomato plant or a pathogenic bacteria, lethal to us? By assembling core genomes and subsequent pan-genomes of life's subgroups, the veil will fall on the similarities and differences (Juhas, M.Trends in Cell Biology 2011). Core genomes are assembled from sequenced genome databases by comparing them to each other and extracting the sequences that remain unchanged, as well as mapping the regions that show slight maleability.(Mar√ßais, G. et al. MUMmer4: A fast and versatile genome alignment system. PLOS Computational Biology 14, e1005944 2018).
------------------------------------------------------------------------------------------------
Scientific Hypothesis: { 2 points } Regions necessary for life will be conserved across all major groups of life, which will allow for a precise isolation of a core "life genome" by simply comparing multiple genomes from across kingdoms of life.

Here, a multiple sequence alignment will be performed comparing a set database of genomes to then extract the ones that overlap across all groups. For each kingdom, k means clustering will group gene sets that make that set different from the generated core genome of life, which will be used as reference. Variation plots will be made for the core genome of all the kingdoms together as well as individually. Furthermore, I will perform a BLAST search to find the current organism that most resembles my core genome assembly to verify its phylogeny, which will act as a control for my methods.
------------------------------------------------------------------------------------------------
Below, this bit of code is responsible for loading in all of the necessary packages. Here is where I will define the packages and what they are loaded in for {3 points for adequately describing the packages}:

1) BioConductor
2) msa :this is the multiple sequence alignment package within bioconductor which has 3 analysis algorithms (ClustalW, ClustalOmega and Muscle)
3) Biostrings is to read the fasta files you want to compare.
4)ggmsa for, as we say in the ed-biz, visual aid
Before running the code below, in the console I needed to install the packages using the commands:

{2 points to write here what commands needed to be typed into the console to install the necessary packages}

```{r}


library(BiocManager)
library(ggplot2)
library(stringi)
library(Biostrings)

library(ggmsa)
library(devtools)

#Now, msa may be faster but I'm having trouble with a missing set of packages called seqmagick and XVector
#library(msa)

#I am also installing multiple sequence alignment visualization tools by the Yu lab at Southern med that use ggplot2, you need to install devtools for it
#if (!requireNamespace("devtools", quietly=TRUE))
   # install.packages("devtools")
#devtools::install_github("YuLab-SMU/ggmsa")
#install.packages('ggmsa')


#to test installation
#library(msa)


# 3 points for loading all of the packages needed correctly
```

-------------------------------------------
Now, I need to load my fasta files for my assorted genomes, which is done below. For this analysis I chose to comapare Arabidopsis, Drosophila, e.Coli, human and S. Cerevisiae. I decided to rather download the available packages in Bioconductor from UCSC straight onto R rather than pull it from a filepath because it was having issues
```{r}

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
install.packages("seqmagick")

BiocManager::install("BSgenome.Celegans.UCSC.ce11")
BiocManager::install("BSgenome.Dmelanogaster.UCSC.dm6")
BiocManager::install("BSgenome.Athaliana.TAIR.TAIR9")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
BiocManager::install("BSgenome.Scerevisiae.UCSC.sacCer3")
BiocManager::install("BSgenome.Ecoli.NCBI.20080805")
```
All of these files were found on the Bioconductor site and they are compiled by the wonderful people at UCSC.
----------------------------------------------------------------------------
```{r}
ggmsa(protein_sequences, 221, 280, seq_name = TRUE, char_width = 0.5) + geom_seqlogo(color = "Chemistry_AA") + geom_msaBar()
```

This code now handles the alignments and the vis-aid. The code aligns the amino acid to allow for the degeneracy of nucleic acid to amino acid code, seeing as AAs are more conserved.This alignment is therefore truer. It is also in charge of building the phylogenetic tree, sequence alignment graph and gene conservation graph that will aid in the core genome development
 
```{r}


#tree
test <- readAAStringSet()
d <- as.dist(stringDist(tp53, method = "hamming")/width(tp53)[1])
tree <- bionj(d)
p_test <- ggtree(tree, branch.length = 'none') + geom_tiplab()
#msa
data_test <- tidy_msa(test)

#gene maps
test_arrow <- readxl::read_xlsx(test)
test_arrow$direction <- 1
test_arrow[test_arrow$strand == "reverse","direction"] <- -1

#color
library(RColorBrewer)
mapping = aes(xmin = start, xmax = end, fill = gene, forward = direction)
my_pal <- colorRampPalette(rev(brewer.pal(n = 10, name = "Set3")))

#tree + gene maps + msa
library(ggnewscale)
p_test + geom_facet(geom = geom_msa, data = data_test,
                     panel = 'msa', font = NULL,
                     border = NA) + xlim_tree(3.5) + 
          new_scale_fill() +
          scale_fill_manual(values = my_pal(10)) +
          geom_facet(geom = geom_motif,
                     mapping = mapping, data = test_arrow,
                     panel = 'genes',  on = 'Test',
                     arrowhead_height = unit(3, "mm"),
                     arrowhead_width = unit(1, "mm"))
```

Now this code below presents the alignments broken down in multiple rows, making visualization of the data a bit easier but you might miss on the patternings (for fun cases like TADs or other spatially determined pattern)

```{r}
#the facet_msa is used to break down the full alignment and field determines the length to be visualized
 ggmsa(protein_sequences, start = 0, end = , font = NULL, color = "Chemistry_AA") + facet_msa(field = 100)
```
 

 Analysis of results shows that
 
 